guide/src/dev/performance.md

```diff
@@ -0,0 +1,206 @@
+}
+
+pub(crate) struct Node<T> {
+    pub child: Option<TreeIndex>,
+    pub next: Option<TreeIndex>,
+    pub item: T,
+}
+```
+
+Key optimizations in the tree structure include:
+
+- Using indices instead of pointers for node references
+- Maintaining a "spine" for fast access to ancestor nodes
+- Storing nodes contiguously in a vector for better cache usage
+- Using non-zero indices to save space in option types
+
+## Single-Pass Parsing
+
+The parser uses a single pass over the input text for both block and inline parsing, avoiding the need to rescan text multiple times. This is achieved through:
+
+1. Maintaining parser state to track context
+2. Handling both block and inline elements incrementally
+3. Using lookahead only when necessary
+
+For example, here's how link reference definitions are handled in a single pass:
+
+```rust
+fn parse_refdef_total(&mut self, start: usize) -> Option<(usize, LinkLabel<'a>, LinkDef<'a>)> {
+    let bytes = &self.text.as_bytes()[start..];
+    if scan_ch(bytes, b'[') == 0 {
+        return None;
+    }
+    let (mut i, label) = self.parse_refdef_label(start + 1)?;
+    i += 1;
+    if scan_ch(&bytes[i..], b':') == 0 {
+        return None;
+    }
+    i += 1;
+    let (bytecount, link_def) = self.scan_refdef(start, start + i)?;
+    Some((bytecount + i, UniCase::new(label), link_def))
+}
+```
+
+## Lazy Text Merging
+
+The library includes a `TextMergeStream` utility that lazily merges consecutive text events:
+
+```rust
+impl<'a, I> Iterator for TextMergeStream<'a, I>
+where
+    I: Iterator<Item = Event<'a>>,
+{
+    type Item = Event<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match (self.last_event.take(), self.iter.next()) {
+            (Some(Event::Text(last)), Some(Event::Text(next))) => {
+                let mut merged = String::from(last);
+                merged.push_str(&next);
+                // Continue merging consecutive text events...
+            }
+            // Handle other cases...
+        }
+    }
+}
+```
+
+This avoids unnecessary string allocations and concatenations when processing text nodes.
+
+## Protection Against Pathological Input
+
+Several protections are in place to prevent exponential time or memory usage on malicious input:
+
+1. Link nesting depth is limited:
+```rust
+pub(crate) const LINK_MAX_NESTED_PARENS: usize = 32;
+```
+
+2. Table column expansion is bounded:
+```rust
+// Limit to prevent quadratic growth from empty cells
+const MAX_AUTOCOMPLETED_CELLS: usize = 1 << 18;
+```
+
+3. Link reference expansion tracking:
+```rust
+// Track expansion to prevent quadratic growth from reference definitions
+let mut link_ref_expansion_limit: usize = text.len().max(100_000);
+```
+
+## Memory Usage Optimization
+
+Several techniques are used to minimize memory usage:
+
+1. Using non-zero indices to save space in option types:
+```rust
+pub(crate) struct TreeIndex(NonZeroUsize);
+```
+
+2. Reusing allocations where possible:
```

Comment:
I don't think `cows` and `links` are really reused. The main purpose of these is to keep tree nodes small and lifetime-free (index over pointer). The latter is not really necessary, `Tree<Item<'input>>` would work too. Some other fields in `Allocations` are reused I suppose.